app/src/org/commcare/tasks/EntityLoaderHelper.kt

```diff
@@ -54,12 +55,23 @@ class EntityLoaderHelper(
         nodeset: TreeReference,
         progressListener: EntityLoadingProgressListener
     ): Pair<List<Entity<TreeReference>>, List<TreeReference>>? {
-        val references = factory.expandReferenceList(nodeset)
-        val entities = loadEntitiesWithReferences(references, progressListener)
-        entities?.let {
-            factory.prepareEntities(entities)
-            factory.printAndClearTraces("build")
-            return Pair<List<Entity<TreeReference>>, List<TreeReference>>(entities, references)
+        if (factory !is AsyncNodeEntityFactory) {
+            // if we are into synchronous mode, cancel background cache work for now to not lock the user db
+            CommCareApplication.instance().currentApp.primeEntityCacheHelper.cancelWork()
```

Comment:
Two questions:
1.  Given that `cancelUniqueWork()` is async, are we certain that while the cancellation is not complete, there won't be any db locking issues?
2.  It seems that calling `cancelLoading()` in sync mode leads to its default implementation which throws an `RuntimeException`. Is that expected?

Reply:
1. Db might block for a little bit (probably a fraction of second) while the cancellation completes but should be able to get the lock as soon as the canncelled thread interrupts and release the lock. 
2. Yeah that's expected, as we only call this method for it's Async implementations.